src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -95651,6 +95648,241 @@ class Box<T>
+                    //             Console.WriteLine(value.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "value").WithLocation(16, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_04()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length == 2 && y?.Length == 2)
+        {
+            Console.WriteLine(x.Length);
+            Console.WriteLine(y.Length);
+        }
+        else
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length); // 2
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (17,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(17, 31),
+                    // (18,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(18, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_05()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
+        string? y = ""43"";
+        if (x?.Length != 2 || y?.Length != 2)
+        {
+            Console.WriteLine(x.Length); // 1
+            Console.WriteLine(y.Length); // 2
+        }
+        else
+        {
+            Console.WriteLine(x.Length);
+            Console.WriteLine(y.Length);
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(x.Length); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 31),
+                    // (13,31): warning CS8602: Dereference of a possibly null reference.
+                    //             Console.WriteLine(y.Length); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(13, 31));
+        }
+
+        [Fact, WorkItem(34942, "https://github.com/dotnet/roslyn/issues/34942")]
+        public void ConditionalAccess_06()
+        {
+            var source =
+@"#nullable enable
+using System;
+
+public class Program
+{
+    static void Main()
+    {
+        string? x = ""42"";
```

Comment:
>""42""; [](start = 20, length = 7)

We should start those tests with a null-state of maybe-null. Otherwise the comparisons with non-null values are no-ops (they update the non-null state to non-null).

Reply:
Changing these to be parameters.